### PR TITLE
Replace NODE API with postgres and update the ETL process

### PR DESCRIPTION
**Pull request by kentpmckinney**
To create a preview app

### CODE REVIEW

File: config/nodeKeys.js [hunk 1]
```diff
@@ -4,11 +4,16 @@ if (process.env.NODE_ENV === "production") {
     NODE_PACKAGE_ID: "e9c55b2c-4019-463e-8efa-622f23221402",
     NODE_LISTING_RESOURCE: "9be4623a-3c01-4b2d-9c7b-567a41abbc1c",
     NODE_PHONE_RESOURCE: "2f66ad5d-5066-49d8-846a-6751cfd23863",
+    /* Heroku creates an environment variable called DATABASE_URL for postgres which is subject to change at any time by Heroku */
+    /* This means we need to use the DATABASE_URL environment variable rather than hard-coding the connection string */
+    PG_CONNECTION_STRING: process.env.DATABASE_URL
```
Comment (timhitchins) [inline config/nodeKeys.js:9]: Would be good to include where to find this in the wiki/readme.

File: client/src/components/App.js [hunk 2]
```diff
@@ -73,77 +74,81 @@ class App extends React.Component {
...
+                          {...props}
+                          records={records}
+                          handleCardSave={this.handleCardSave}
+                          savedDataId={savedDataId}
+                        />
+                      )}
+                    />
+                    <Route
+                      exact path="/admin" render={() =>
+                        window.location.href = [window.location.protocol, '//', window.location.host.replace(/\d+/, '5000'), '/admin/dashboard'].join('')
```
Comment (timhitchins) [inline client/src/components/App.js:141]: It would probably be good to investigate using React Router's history props here instead of the window object.

File: ETL/main.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    lon = ''
+    try:
+        lat = location.latitude
+    except:
+        lat = ''
+    try:
+        lon = location.longitude
+    except:
+        lon = ''
+    sql = "UPDATE etl_staging_1 SET lat='{0}', lon='{1}' WHERE id='{2}';".format(
```
Comment (timhitchins) [inline ETL/main.py:163]: Personally a fan of f-strings but that is just a suggestion.

File: ETL/main.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+cursor.execute(query)
+db.commit()
+
+# --------- PHASE 3: Geocode addresses --------- #
+
+log("Geocode addresses")
+geolocator = Nominatim(user_agent="rose-city-resource")
+query = "SELECT * FROM etl_staging_1;"
+cursor.execute(query)
+columns = []
```
Comment (timhitchins) [inline ETL/main.py:140]: Would be good to leave some space between each loop logic for increased readability.

File: client/src/components/Details.js [hunk 5]
```diff
@@ -15,9 +15,10 @@ class Details extends React.Component {
   state = { cardDetailsData: null };
 
   componentDidMount() {
-    const { nodeData } = this.props;
+    const { records } = this.props;
     const queryVals = queryString.parse(this.props.location.search);
```
Comment (timhitchins) [inline client/src/components/Details.js:19]: Here is an example of React Router's location prop fyi.

File: index.js [hunk 6]
```diff
@@ -1,34 +1,61 @@
...
+require("./routes/admin")(app, pool);
 
-//production boilerplate
+/* Check for database connectivity and provide a human-friendly message on failure */
+const testDatabaseQuery = () => {
+  pool.query(`select last_update from production_meta`, (err, res) => {
+    if (err) {
+      console.log('Error connnecting to the database!');
+      if (process.env.DATABASE_URL === undefined || process.env.DATABASE_URL === null || process.env.DATABASE_URL === '') {
+        console.log('Please check that the DATABASE_URL environment variable is correct');
```
Comment (timhitchins) [inline index.js:45]: Usually good to use `console.error()` for these situations.

File: config/nodeKeys.js (hunk 1 above)
Comment (timhitchins) [inline config/nodeKeys.js:9]: Might be more work than it is worth, but it could be good to rename this env var to `DATABASE_URI` to keep things consistent with heroku.

File: config/nodeKeys.js [hunk 7]
```diff
@@ -4,11 +4,16 @@ if (process.env.NODE_ENV === "production") {
...
   };
 } else {
   module.exports = {
     NODE_PACKAGE_ID: "592c18db-efa6-44c6-8477-4ffa4103ba94",
     NODE_LISTING_RESOURCE: "61cee891-7d0f-4ebe-b8ea-c0c8d6cb27e7",
     NODE_PHONE_RESOURCE: "4407461b-e99d-4d8e-8a44-18483aa8d13c",
+    /* The local development postgres connection string requires manually adding environment variables */
+    PG_CONNECTION_STRING: process.env.DATABASE_URL
```
Comment (timhitchins) [inline config/nodeKeys.js:17]: @kentpmckinney  Where will this env var be found outside of a dev environment?  You may have already mentioned it, but how does Heroku know about this var?

File: ETL/main.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    sys.exit(11)
+
+DATABASE_URL = os.environ.get('DATABASE_URL')
+
+if DATABASE_URL == None or DATABASE_URL == '':
+    if sys.argv[1] != None and sys.argv[1] != '':
+        DATABASE_URL = sys.argv[1]
+
+if DATABASE_URL == None or DATABASE_URL == '':
+    print('The postgres connection string DATABASE_URL is missing')
```
Comment (timhitchins) [inline ETL/main.py:27]: Could be good to have a better message here like where to find it.

File: ETL/main.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+def create_import_table(table_name, json_data):
+    # Data from airtables is NOT homogenous
+    # Normalize the JSON data to simplify processing
+    data = pandas.json_normalize(json_data)
+
+    # Strip off the 'fields.' prefix that pandas adds to the column names
+    columns = data.columns
+    new_columns = []
+    for column in columns:
+        new_columns.append(str(column).replace(
```
Comment (timhitchins) [inline ETL/main.py:56]: List comprehensions tend to be more modern / readable but gain this is just a suggestion.

File: ETL/main.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+    # Iterate through each row and add values
+    sql = "INSERT INTO {} (".format(table_name) + \
+        ','.join(data.columns) + ") VALUES "
+    allrowvalues = []
+    for index, row in data.iterrows():
+        currentrowvalues = []
+        for col in data.columns:
+            # Get the values for the current row and sanitize them into clean string data
+            currentrowvalues.append(
```
Comment (timhitchins) [inline ETL/main.py:79]: list comprehension here too, but again a sugestion.

File: ETL/main.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+log("Merge the import tables")
+query = ("select etl_merge_import_tables();")
+cursor.execute(query)
+db.commit()
+
+# --------- PHASE 3: Geocode addresses --------- #
+
+log("Geocode addresses")
+geolocator = Nominatim(user_agent="rose-city-resource")
```
Comment (timhitchins) [inline ETL/main.py:137]: Confused as to how this user agent works.

File: ETL/main.py (hunk 3 above)
Comment (timhitchins) [reply ETL/main.py:163]: again just a suggestion.

File: client/src/components/App.js [hunk 12]
```diff
@@ -9,19 +9,20 @@ import Nav from "./Nav";
 import Footer from "./static_components/Footer";
 import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
 import {
-  getNodeData,
-  getPackageData,
+  getRecords,
+  getRecordsLastUpdatedTimestamp,
```
Comment (timhitchins) [inline client/src/components/App.js:13]: nice

File: client/src/components/Results/Cards.js [hunk 13]
```diff
@@ -69,9 +69,9 @@ class Card extends React.Component {
       parsedListing: record.listing,
       parsedPhone: cardPhoneTextFilter(record),
       parsedWeb: cardWebAddressFixer(record.website),
-      parsedStreet: `${cardTextFilter(record.street)} ${cardTextFilter(
+      parsedStreet: record.street !== null && record.street !== '' ? `${cardTextFilter(record.street)} ${cardTextFilter(
```
Comment (timhitchins) [inline client/src/components/Results/Cards.js:72]: Are the db string results different from before?

File: client/src/components/Results/Results.js [hunk 14]
```diff
@@ -186,8 +186,8 @@ export default Results;
 //   const data = await nodeQueryBuilder(dataMap.category, dataMap.parent);
```
Comment (timhitchins) [inline client/src/components/Results/Results.js:186]: We could probably get rid of all this commented code.

File: client/src/components/static_components/Footer.js [hunk 15]
```diff
@@ -23,7 +23,7 @@ const Footer = (props) => {
             </a>
             </address>
           </div>
-          <div className="footer-item node-logo">
+          {/* <div className="footer-item node-logo">
```
Comment (timhitchins) [inline client/src/components/static_components/Footer.js:26]: same here

File: client/src/utils/distance.js [hunk 16]
```diff
@@ -22,16 +22,16 @@ function getCurrentLocation(options) {
...
-  }
-}
+  return await getCurrentLocation({
+    enableHighAccuracy: true,
+    timeout: 5000,
+    maximumAge: 0
+  })
+    .catch(e => {
+      if (e.name === 'PositionError') {
+        console.log(e.message + '. code = ' + e.code);
```
Comment (timhitchins) [inline client/src/utils/distance.js:34]: I probably wrote this one in the past, but would be good to update to a `console.error()` and return a value so we don't break the ui in a position error.

File: config/create.sql [hunk 17]
```diff
@@ -0,0 +1,147 @@
+/* SQL commands to create the tables used by this version of Rose City Resource */
+
+/*
+    NOTE:
```
Comment (timhitchins) [inline config/create.sql:4]: Thank you for this context.

Comment (timhitchins) [reply config/create.sql:4]: We should probably add this info to the wiki/readme.

File: initializePassport.js [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+  };
+
+  passport.use(
+    new LocalStrategy(
+      { usernameField: "email", passwordField: "password", roleField: 'role' },
+      authenticateUser 
+    )
+  );
+
+  // passport.use(new LocalStrategy({
```
Comment (timhitchins) [inline initializePassport.js:41]: probably be good to clean up all this commented out code.

File: initializePassport.js [hunk 19]
```diff
@@ -0,0 +1,76 @@
+const LocalStrategy = require("passport-local").Strategy;
+const bcrypt = require("bcrypt");
+
+async function initialize(passport, pool) {
+
+  const authenticateUser = async (email, password, done) => {
+    await pool.query(
+      `SELECT * FROM production_user WHERE email = $1`, [email],
+      async (err, results) => {
+        if (err) { throw err; }
```
Comment (timhitchins) [inline initializePassport.js:10]: Some better error messaging here would probably be good.

File: package.json [hunk 20]
```diff
@@ -29,10 +29,18 @@
     "nodemon": "^2.0.3"
   },
   "dependencies": {
+    "bcrypt": "^5.0.0",
     "compression": "^1.7.4",
     "cors": "^2.8.5",
+    "ejs": "^3.1.5",
     "express": "^4.17.1",
+    "express-flash": "0.0.2",
```
Comment (timhitchins) [inline package.json:37]: Nice.  Never know bout express-flash.

File: routes/admin.js [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+    }
+  });
+
+  /* API method to pull logs from the public.etl_run_log table */
+  app.get("/admin/dashboard/etl-log", async (req, res, next) => {
+    try {
+
+      let log = null;
+
+      await pool.query('select * from etl_run_log order by time_stamp asc;', async (err, result) => {
```
Comment (timhitchins) [inline routes/admin.js:88]: Would be good to keep query case/syntax consistent.  Here the query is lower case but in other routes it is upper.

File: routes/admin.js [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+    session({
+      secret: 'secret',
+      resave: false,
+      saveUninitialized: false
+    })
+  );
+  app.use(passport.initialize());
+  app.use(passport.session());
+  app.use(flash());
+  app.use(function (req, res, next) {
```
Comment (timhitchins) [inline routes/admin.js:24]: So if I understand correctly this is the middleware to handle errors?

Comment (timhitchins) [reply routes/admin.js:24]: ...just trying to understand.  We should probably make sure that that we have a `res.status(500)` somewhere in the error handling.

File: routes/admin.js [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+      return next();
+    }
+    res.redirect("/admin/dashboard");
+  }
+
+  /* Clear all tables like public.etl_% */
+  const clearTables = async () => {
+    await pool.query('select etl_clear_tables();', async (err, result) => {
+      if (err) {
+        console.log(err)
```
Comment (timhitchins) [inline routes/admin.js:256]: Some more messaging with a `console.error()` would be good for this and the ones below.  `console.error('Error executing query', err.stack)` comes to mind.

File: routes/last-update.js [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+    try {
+
+      /* Pull the listing table and parse into JSON */
+      await pool.query("SELECT last_update FROM production_meta", async (sqlerr, sqlres) => {
+        if (sqlerr) {
+          if (process.env.NODE_ENV == undefined || process.env.NODE_ENV !== "production") {
+            try {
+              await res.send(sqlerr);
+            } catch (e) {
+              console.log(e);
```
Comment (timhitchins) [inline routes/last-update.js:12]: add error handling.

File: views/dashboard.ejs [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+      if (hasStartMessage && !hasEndMessage) {
+        document.getElementById('status').innerText = 'There is already an import in progress. Please allow up to 30 minutes for that process to complete.';
+      }
+
+      /* Trigger the ETL process to start */
+      fetch(window.location.href + '?' + new URLSearchParams({ action: 'runetl' }), { method: 'GET' })
+        .catch(e => { document.getElementById('status').innerText = 'Failed to start import process'; })
+
+      /* Query the server for ETL status on an interval */
+      var stagingInterval = window.setInterval(async function () {
```
Comment (timhitchins) [inline views/dashboard.ejs:121]: Mabe use a `const` here

File: views/dashboard.ejs [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+      let confirmation = confirm("The LIVE SITE will be updated. Proceed?");
+      if (confirmation == false) { return; }
+      document.getElementById('status').innerText = 'Preparing to update production...';
+
+      /* Trigger the production udpate process to start */
+      const status = await fetch(window.location.href + '?' + new URLSearchParams({ action: 'runprod' }), { method: 'GET' })
+        .catch(e => { document.getElementById('status').innerHTML = '<span>&#10008;&nbsp;Failed to start import process</span>'; })
+
+      /* Query the server for status on an interval */
+      var prodInterval = window.setInterval(async function () {
```
Comment (timhitchins) [inline views/dashboard.ejs:159]: const or let

File: routes/admin.js [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+  app.get("/admin/dashboard", userIsAuthenticated, async (req, res, next) => {
+    try {
+
+      const { action } = req.query;
+
+      if (action === 'runetl') {
+        /* The 'Import to Staging' button was clicked */
+
+        /* Prepare to run the ETL script */
+        await clearTables().catch(e => console.log(e));
```
Comment (timhitchins) [inline routes/admin.js:40]: add some more error messaging here with a `console.error()`

File: routes/query.js [hunk 28]
```diff
@@ -0,0 +1,26 @@
+module.exports = (app, pool) => {
```
Comment (timhitchins) [inline routes/query.js:1]: So this is the route for production queries correct?

File: views/dashboard.ejs [hunk 29]
```diff
@@ -0,0 +1,293 @@
...
+        const stagingFinalized = logs.filter(L => L.message === 'Finalize the staging table');
+        if (stagingFinalized && stagingFinalized.length >= 1) {
+          const items = await fetchDataValidationItems();
+          renderValidationItems(items)
+        }
+        if (jobStart && jobStart.length >= 1) {
+          document.getElementById('status').innerHTML = 'Import process is running...';
+        }
+        if (jobEnd && jobEnd.length >= 1) {
+          const previewPage = [window.location.protocol, '//', window.location.host.replace(/\d+/, '3000'), '?datatable=staging'].join('');
```
Comment (timhitchins) [inline views/dashboard.ejs:139]: could be a good opportunity for `query-string` but just a suggestion.

File: views/dashboard.ejs [hunk 30]
```diff
@@ -0,0 +1,293 @@
...
+        }
+      }, 4000)
+
+    }
+
+    async function production() {
+
+      /* Prompt for confirmation */
+      let confirmation = confirm("The LIVE SITE will be updated. Proceed?");
+      if (confirmation == false) { return; }
```
Comment (timhitchins) [inline views/dashboard.ejs:151]: what is the purpose for this empty return?

File: views/dashboard.ejs [hunk 31]
```diff
@@ -0,0 +1,293 @@
...
+  </details>
+  <br>
+  <details open>
+    <summary>Logs</summary>
+    <table cellpadding='0' cellspacing='0'
+      style='margin-left:auto; margin-right:auto; font-size:16px; font-family:monospace; border-spacing: 10px 0;'>
+      <tbody id='etl-log'></tbody>
+    </table>
+  </details>
+  <script>
```
Comment (timhitchins) [inline views/dashboard.ejs:252]: cool stuff

File: views/dashboard.ejs [hunk 32]
```diff
@@ -0,0 +1,293 @@
+<head>
+  <title>Rose City Resource Admin Page</title>
+  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/skeleton/2.0.4/skeleton.css" />
+  <script>
+
```
Comment (timhitchins) [inline views/dashboard.ejs:5]: Might be good to separate out this js your html template for better readability in this file.

File: views/dashboard.ejs [hunk 33]
```diff
@@ -0,0 +1,293 @@
...
+    async function fetchPgRows() {
+      let data = await fetch(window.location.href + '/pg-rows')
+        .catch(e => console.log(e))
```
Comment (timhitchins) [inline views/dashboard.ejs:8]: same comments about error handling.  Might be a good function to create so it can be reused in the catch blocks below.

**Review by timhitchins — CHANGES_REQUESTED** [review]
Hey @kentpmckinney and @wrunion great job on getting this implemented!!  Thank you!  I left a somewhat thorough review regarding error handling details and general things that I noticed that would improve consistency.   Let me know if you have questions.  I would also love to have one more meeting if you all are available so we can discuss our merge into production plan. I also plan to submit a couple PRs over then next couple weeks that will improve some of the fe mess I originally created.

File: routes/admin.js (hunk 22 above)
Comment (wrunion) [reply routes/admin.js:24]: Man, github doesn't make it easy to see what code is being commented on! Took me a hot second to find and see the complete code :) 

So, the answer is both yes and no. IIUC, express-flash is specifically for showing messages to the user, based on their input, and doesn't map to (or handle) server-side errors at all. So we do definitely need some true error handling! 

For context, the express-flash messages function a lot like input error messages you'd get in an html form in the browser; Examples of flash ""error"" messages would be "that email isn't registered," "passwords do not match" etc. They're not true errors from the code side of things, just mistakes the user needs to fix. 

Overall, I definitely hear you about (actual code) error-handling! That's definitely something we need to fix. I'll put that on the agenda for the meeting today, to figure out how we want to do that. 

tl;dr - express-flash messages are for display only, to the user. They can be buggy in and of themselves, and aren't totally necessary, so we can think about either getting rid of them, or making sure we iron out the kinks before going live. And we definitely need better error handling in general, so let's talk about the best way to do that :)

File: ETL/main.py (hunk 8 above)
Comment (kentpmckinney) [reply ETL/main.py:27]: Added more detail to the error message

File: ETL/main.py (hunk 11 above)
Comment (kentpmckinney) [reply ETL/main.py:137]: It passes the value as the User-Agent when the library makes HTTP requests. The usage policy for the library requires setting it to the name of the application.

Comment (kentpmckinney) [reply ETL/main.py:137]: Added a comment to the Python script

File: ETL/main.py (hunk 4 above)
Comment (kentpmckinney) [reply ETL/main.py:140]: Added blank lines for readability

File: client/src/components/Results/Cards.js (hunk 13 above)
Comment (kentpmckinney) [reply client/src/components/Results/Cards.js:72]: Yes

File: client/src/components/Results/Results.js (hunk 14 above)
Comment (kentpmckinney) [reply client/src/components/Results/Results.js:186]: Done

File: client/src/components/static_components/Footer.js (hunk 15 above)
Comment (kentpmckinney) [reply client/src/components/static_components/Footer.js:26]: Done

File: client/src/utils/distance.js (hunk 16 above)
Comment (kentpmckinney) [reply client/src/utils/distance.js:34]: Done.

File: config/create.sql (hunk 17 above)
Comment (kentpmckinney) [reply config/create.sql:4]: Done

File: config/nodeKeys.js (hunk 1 above)
Comment (kentpmckinney) [reply config/nodeKeys.js:9]: Added

Comment (kentpmckinney) [reply config/nodeKeys.js:9]: Done, but renamed to DATABASE_URL to match the name verbatim that Heroku uses

File: config/nodeKeys.js (hunk 7 above)
Comment (kentpmckinney) [reply config/nodeKeys.js:17]: There are comments in the file which explain that Heroku automatically creates and manages an environment variable with this name. When Heroku performs maintenance, for example, Heroku will update DATABASE_URL and restart the application.

File: index.js (hunk 6 above)
Comment (kentpmckinney) [reply index.js:45]: I have gone through the code and replaced console.log with console.error

File: routes/admin.js (hunk 27 above)
Comment (kentpmckinney) [reply routes/admin.js:40]: Done

File: routes/admin.js (hunk 21 above)
Comment (kentpmckinney) [reply routes/admin.js:88]: Fixed

File: routes/admin.js (hunk 23 above)
Comment (kentpmckinney) [reply routes/admin.js:284]: Done

File: routes/last-update.js (hunk 24 above)
Comment (kentpmckinney) [reply routes/last-update.js:12]: I added status 500 handling for these

File: routes/query.js (hunk 28 above)
Comment (kentpmckinney) [reply routes/query.js:1]: Yes, it pulls the 'production' or 'live site' set of data. query-staging pulls the 'preview' data. The terminology is confusing, but the concept is simple. The 'production' site can use 'preview' DATA, but here 'preview' refers only to the data itself.

File: views/dashboard.ejs (hunk 25 above)
Comment (kentpmckinney) [reply views/dashboard.ejs:121]: Added

File: views/dashboard.ejs (hunk 30 above)
Comment (kentpmckinney) [reply views/dashboard.ejs:151]: Just for edification, to confirm intent that nothing is meant to be returned.

File: views/dashboard.ejs (hunk 26 above)
Comment (kentpmckinney) [reply views/dashboard.ejs:159]: Done

File: ETL/main.py (hunk 11 above)
Comment (timhitchins) [reply ETL/main.py:146]: Awesome thank you!

File: routes/admin.js (hunk 22 above)
Comment (timhitchins) [reply routes/admin.js:24]: thank your for the context.

File: ETL/main.py (hunk 9 above)
Comment (kentpmckinney) [reply ETL/main.py:60]: So you mean something like this: columns = [str(column).replace("fields.", "").replace(" ", "_").replace("-", "")) for column in data.columns]?

File: ETL/main.py (hunk 3 above)
Comment (kentpmckinney) [reply ETL/main.py:177]: Thanks! I didn't know about f-strings, and now I'm gonna use 'em all the time! Updated the script.

File: client/src/components/App.js (hunk 2 above)
Comment (kentpmckinney) [reply client/src/components/App.js:141]: I tried this, and for whatever reason the React Router props don't seem to include the host name or port, which is all this bit of code needs. Instead, they only seem to contain the relative path and query string. I searched on Stack Overflow for alternatives to using window.location, because it's an eyesore to me as well, but I wasn't able to find anything better. Apparently, props.location does work for some folks and not for others. Maybe newer versions of React only show the relative path? I see a lot of people on that site using window.location with React without any lambasting from their peers, which nearly validates it. I want to make this code better if it is possible, but I don't see any way to do it.

File: client/src/components/Details.js (hunk 5 above)
Comment (kentpmckinney) [reply client/src/components/Details.js:19]: See above comment

File: views/dashboard.ejs (hunk 29 above)
Comment (kentpmckinney) [reply views/dashboard.ejs:139]: This code is tricky, since it is meant to work both for local and hosted environments. It needs to replace the port number if one exists. query-string just parses a given URL, if I'm looking at the documentation correctly.

File: routes/query.js (hunk 28 above)
Comment (timhitchins) [reply routes/query.js:1]: Awesome thank you for explaining!

File: views/dashboard.ejs (hunk 29 above)
Comment (timhitchins) [reply views/dashboard.ejs:139]: @kentpmckinney Yes you are totally right.  Hope I didn't send you down a rabbit hole.

**Review by timhitchins — APPROVED** [review]
Hey @kentpmckinney and @wrunion.  I just approved this to keep from blocking it. We should probably just go over all the most recent changes (including the wiki) since this review was started before merging though.
